# Imports

In [110]:
import pandas as pd

# read csv

In [111]:
cancer_df = pd.read_csv("CANCER_TABLE.csv")

# Inspect data

In [112]:
cancer_df.head(5)

,diameter (cm),cancer
0,6.309685,False
1,4.580894,False
2,6.310071,False
3,7.495139,False
4,2.216181,False


In [113]:
cancer_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   diameter (cm)  1000 non-null   float64
 1    cancer        1000 non-null   bool   
dtypes: bool(1), float64(1)
memory usage: 8.9 KB


In [114]:
for col in cancer_df.columns:
    print(repr(col))

'diameter (cm)'
' cancer'


THERE is a space in the name of cancer -_-

In [115]:
cancer_df.columns = cancer_df.columns.str.strip()

In [116]:
cancer_df["cancer"].value_counts()

cancer
False    807
True     193
Name: count, dtype: int64

In [117]:
cancer_df.rename(columns={"diameter (cm)": "diameter"}, inplace=True)

In [118]:
cancer_df.describe()

,diameter
count,1000.000000
mean,4.971051
std,2.934147
min,0.059294
25%,2.401131
50%,4.923567
75%,7.605594
max,9.994688


Lets create thier model

In [119]:
model = lambda diameter: 1 if diameter > 7 else 0

In [120]:
cancer_df["prediction"] = cancer_df.diameter.apply(model)

TP:

In [121]:
true_positive = cancer_df[
    (cancer_df.prediction == True) & 
    (cancer_df.cancer == True)].prediction.count()
true_positive

np.int64(193)

193 true positive
now lets do TN

TN

In [122]:
true_negative = cancer_df[
    (cancer_df.prediction == False) & 
    (cancer_df.cancer == False)].prediction.count()
true_negative

np.int64(689)

689 True Negative

FP

In [123]:
false_positive = cancer_df[
    (cancer_df.prediction == True) & 
    (cancer_df.cancer == False)].prediction.count()
false_positive

np.int64(118)

118 FP

FN

In [124]:
false_negative = cancer_df[
    (cancer_df.prediction == False) & 
    (cancer_df.cancer == True)].prediction.count()
false_negative

np.int64(0)

0 FN
means that when we guess someone doesnt have cancer he is forsure doesnt have cancer
thats probably because only ~25-30 percent have higher than 7 diameter
so if we guess on almost all of them as negative we can cover most if not all of the negatives (in this case all)

now using all that lets calcualte some metrics

Recall / True Positive Rate

In [125]:
recall = true_positive / (false_negative + true_positive)
recall*100

np.float64(100.0)

sooooo out of all people who actually developed cancer (false_negative + true_positive) what percentage did we catch?
we caught all of them.

FPR (False Positive Rate)

In [126]:
false_positive_rate = false_positive / (false_positive + true_negative)
false_positive_rate*100

np.float64(14.622057001239158)

out of all the negative people (those who ddint have cancer) what percentage did we incorrectly flagged?
14.62% of people who did not develop cancer were incorrectly flagged as cancer-positive.

Precision

In [127]:
precision = true_positive / (true_positive + false_positive)
precision*100

np.float64(62.057877813504824)

Of everyone predicted to develop cancer, 62.06% actually developed cancer.

Accuracy

In [128]:
accuracy = (true_positive + 
            true_negative) / (true_positive + 
                              true_negative + 
                              false_positive + 
                              false_negative)
accuracy*100

np.float64(88.2)

what fraction of all predictions were correct
damn we are quite good (but the dataset aint that spread out there are too many non-cancer patients)

F1 Score

In [129]:
f1 = (2 * precision * recall) / (precision + recall)
f1*100

np.float64(76.5873015873016)

high balance between the recall and precision (although precision is limited by false positives...)